### ISSUE TITLE
Better compiler diagnostics (error messages, warnings, hints, etc.)

### ISSUE THREAD

**Issue by anton-trunov** [tact-lang/tact#183]
There is a lot to be done in this area. I would say we should target the level of diagnostic achieved by Rust.

Also, here are some excellent references, collected by @ironcev while working on [this RFC](https://github.com/FuelLabs/sway-rfcs/blob/master/rfcs/0011-expressive-diagnostics.md) for the [Sway](https://github.com/FuelLabs/sway) smart contract language (thanks Igor <3).

- [Expressive Diagnostics](https://clang.llvm.org/diagnostics.html)
- [Shape of errors to come](https://blog.rust-lang.org/2016/08/10/Shape-of-errors-to-come.html)
- [Compiler Errors for Humans](https://elm-lang.org/news/compiler-errors-for-humans)
- [Compilers as Assistants](https://elm-lang.org/news/compilers-as-assistants)
- [Concepts Error Messages for Humans](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2429r0.pdf)
- [Compiler Error Messages Considered Unhelpful: The Landscape of Text-Based Programming Error Message Research](https://web.eecs.umich.edu/~akamil/papers/iticse19.pdf)
- [Usability improvements in GCC 8](https://developers.redhat.com/blog/2018/03/15/gcc-8-usability-improvements#)

Some more:
- https://github.com/johnyob/grace -- an OCaml 🐪 library that includes a series of interfaces for building, reporting, and rendering beautiful compiler errors 📜
- https://github.com/zesterer/ariadne — A fancy diagnostics & error reporting Rust crate

**Comment by novusnota** [comment]
If anything, it would be nice to have unified structure for error messages, such that they can be reliably parsed in tools, such as tact.vim, tact-vscode or linters & such.

Maybe, as a first step on this path Tact should make either a programmatic (`import {...} from '@tact-lang/compiler'`) or a CLI (`tact --...`) way to output all compilation issues (syntax, errors, etc.) in some machine-readable format (such as JSON, for example). That would remove the current need in parsing error messages by hand trying to match the moving target (#182), and thus drastically simplify the work for tool makers.

And once that done, proceed think about making unified and user-friendly messages for the stdout.

cc @anton-trunov @byakuren-hijiri @Gusarich what do you think?

**Comment by anton-trunov** [comment]
@novusnota Yes, having a centralized data structure for all error messages is what we should go for

**Comment by byakuren-hijiri** [comment]
@novusnota I agree that it would be great to have a structure designed to describe errors.
This structure could be formatted for readability by humans or serialized into JSON to make output accessible for tools.

I think its format should be as simple as possible, since standard formats like SARIF seem too complicated.

**Comment by novusnota** [comment]
A good addition to that list:
* [The Syntax Cliff](https://elm-lang.org/news/the-syntax-cliff), from Elm language blog — discusses how to make syntax error messages so nice, that they're actually tutoring the language to people!
* [Error Message Catalog of Elm](https://github.com/elm/error-message-catalog/issues), where is a bunch of suggestions and implemented fixes (see closed issues) on error messages: their clarity, readability, preciseness.

There's another language, which also excels at developer experience — [Gleam](https://gleam.run). I didn't see the them do similar discussions on error messages in their blog, though they are very concerned about good tools and convenience of using them and the language — take a look at [their 1.0 release package](https://gleam.run/news/gleam-version-1/): fledged language design, compiler, build tool, formatter, package manager, language server (LSP spec), and various bindings to WASM/JS. And all those are built-in the compiler binary! Plus, an [interactive tour](https://tour.gleam.run ), like seen in [Golang](https://go.dev/tour/welcome/1) and [Elm](https://elm-lang.org/try). Btw, it would be nice to make something similar in Nujan, although we have cool https://tact-by-example.org already.

Gleam has a bunch of other tools too, I'll post them here for completeness (and later reference):
* [VSCode extension](https://github.com/gleam-lang/vscode-gleam) (as a wrapper around the language server binary)
* [Tree-sitter grammar](https://github.com/gleam-lang/tree-sitter-gleam)
* [Vim support](https://github.com/gleam-lang/gleam.vim)
* [Emacs support](https://github.com/gleam-lang/gleam-mode)
* And unofficial [Sublime Text support](https://github.com/digitalcora/sublime-text-gleam)

**Comment by novusnota** [comment]
> If anything, it would be nice to have unified structure for error messages...

Suggestion: Start with parsing Ohm's syntax error messages, instead of simply adding additional info on top ([grammar.ts](https://github.com/tact-lang/tact/blob/main/src/grammar/grammar.ts)). That way we'll have a structure for representing syntax errors inside the compiler, which may later get re-used for semantical errors, build errors, warnings or hints.

Because at the moment,output of `check()` from `@tact-lang/compiler` is not much different from Ohm's output, which isn't helpful for the programmer (at least in the following example):

![image](https://github.com/tact-lang/tact/assets/68142933/d0063c17-ff0f-4c94-8b4a-8c6bf07ee4a4)

P.S.: The syntax error with `internal()` instead of `receive()` is done intentionally :)
